runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt

```diff
@@ -68,8 +56,13 @@ private sealed class AbstractJsonTreeOutput(
     }
 
     override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
-        encodePolymorphically(serializer, value) {
-            writePolymorphic = true
+        if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
```

Comment:
Please leave a short comment why this check is here (especially `currentTagOrNull`)